fairscale/nn/data_parallel/offload_ddp.py

```diff
@@ -0,0 +1,402 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+A distributed data parallel class that shards the model and the optimizer into pieces.
+
+See https://github.com/pytorch/pytorch/issues/42849 for more context. Credits to Shen Li for the original idea
+
+"""
+
+from builtins import isinstance
+import enum
+import logging
+from typing import Any, Dict, List, Optional, Type
+
+import torch
+from torch import nn
+import torch.distributed as dist
+
+
+class SplitStrategy(enum.Enum):
+    """
+    Strategy used to shard a model. Each shard can hold the same number of layers, or hold the same cumulated parameter size.
+    """
+
+    NUM_LAYERS = enum.auto()
+    MEMORY = enum.auto()
+
+
+def _split(modules: nn.Sequential, number_shards: int, strategy: SplitStrategy) -> List[List[nn.Module]]:
```

Comment:
@min-xu-ai This decides how to shard the model, by default it counts the cumulated parameter size per shard and equalizes that